hasher-matcher-actioner/hmalib/models.py

```diff
@@ -16,14 +19,96 @@
 """
 
 
-class DynamoDBItem:
+class DynamoDBCountMixin:
+    """
+    This is a mixin for the base class (DynamoDBItem). It is not required to be
+    mixed in to the concrete classes. Use this just to group common
+    functionality together.
+
+    Most methods defined are class methods and not static methods. Static
+    methods are unbounded which makes it impossible to override and refer to in
+    sub classes.
+    """
+
+    @classmethod
+    def is_total_countable(cls):
+        """
+        To be declared by sub classes. If true, a counter record is maintained
+        in the same table. Be advised, that the counter is incremented with
+        every 'write' be it a create or update.
+        """
+        return False
+
+    @classmethod
+    def get_total_count_keyname(cls):
+        """
+        Is used to store and retrieve count values. To be implemented by
+        concrete classes alone. Unless is_total_countable() is True, this method
+        need not be implemented.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def get_total_count(cls, table: Table) -> t.Optional[int]:
+        """
+        See `is_total_countable`. If self.is_total_countable is False, return
+        None. Else return the number of writes this table has seen.
+        """
+        if cls.is_total_countable():
+            response = table.get_item(
+                Key={"PK": cls._get_count_pkey(), "SK": cls._get_count_skey()}
+            )
+            return (
+                "Item" in response
+                and int(t.cast(Decimal, response["Item"]["WriteCount"]))
+                or 0
+            )
+        else:
+            return None
+
+    @classmethod
+    def _get_count_pkey(cls) -> str:
+        return "counts"
+
+    @classmethod
+    def _get_count_skey(cls) -> str:
+        return f"count#{cls.get_total_count_keyname()}"
+
+
+class DynamoDBItem(DynamoDBCountMixin):
 
     CONTENT_KEY_PREFIX = "c#"
     SIGNAL_KEY_PREFIX = "s#"
     TYPE_PREFIX = "type#"
 
     def write_to_table(self, table: Table):
-        table.put_item(Item=self.to_dynamodb_item())
+        client: Client = table.meta.client
+        transact_items: t.List[TransactWriteItemTypeDef] = [
+            {
+                "Put": {
+                    "Item": self.to_dynamodb_item(),
+                    "TableName": table.table_name,
+                },
+            },
+        ]
+
+        if self.__class__.is_total_countable():
```

Comment:
question: is `__class__` needed here for just defensive programing? (want to check my understanding of inheritance and classmethods)

Reply:
Just being explicit. You could just as easily do `self.is_total_countable()`. Nothing in python prevents this.